### PR TITLE
L3out bgp protocol profile

### LINKED ISSUE #416: aci_l3out_bgp_protocol_profile: Support for bgpRsBestPathCtrlPol
We should add the support for the `bgpRsBestPathCtrlPol` policy like we do for the timers.

### PR DESCRIPTION
**Pull request by xinyuezhao**
add relationship for l3out bgp protocol profile
fix attributes for bgp protocol profile
update example for bgp prop
update aci-go-client as well

resolve #416

### CODE REVIEW

File: aci/resource_aci_bgpprotp.go [hunk 1]
```diff
@@ -36,13 +36,23 @@ func resourceAciL3outBGPProtocolProfile() *schema.Resource {
 				Optional: true,
 				Computed: true,
 			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Optional: true,
+				Computed: true,
+			},
```
Comment (lhercot) [inline aci/resource_aci_bgpprotp.go:44]: do we need this empty line?

File: aci/resource_aci_bgpprotp.go [hunk 2]
```diff
@@ -36,13 +36,23 @@ func resourceAciL3outBGPProtocolProfile() *schema.Resource {
...
 
 			"annotation": &schema.Schema{
 				Type:     schema.TypeString,
 				Optional: true,
 				Default:  "orchestrator:terraform",
 			},
```
Comment (lhercot) [inline aci/resource_aci_bgpprotp.go:50]: do we need this empty line?

File: aci/resource_aci_bgpprotp.go [hunk 3]
```diff
@@ -138,49 +162,67 @@ func resourceAciL3outBGPProtocolProfileCreate(ctx context.Context, d *schema.Res
 	}
 	d.Partial(false)
 
+	if relationTobgpRsBestPathCtrlPol, ok := d.GetOk("relation_bgp_rs_best_path_ctrl_pol"); ok {
+		relationParam := relationTobgpRsBestPathCtrlPol.(string)
```
Comment (lhercot) [inline aci/resource_aci_bgpprotp.go:166]: Maybe put the GetMoName() here like the next function do.

File: aci/resource_aci_bgpprotp.go [hunk 4]
```diff
@@ -226,32 +276,47 @@ func resourceAciL3outBGPProtocolProfileRead(ctx context.Context, d *schema.Resou
...
 	}
+
+	bgpRsBestPathCtrlPolData, err := aciClient.ReadRelationbgpRsBestPathCtrlPol(dn)
+	if err != nil {
+		log.Printf("[DEBUG] Error while reading relation bgpRsBestPathCtrlPol %v", err)
+		d.Set("relation_bgp_rs_best_path_ctrl_pol", "")
+	} else {
+		if _, ok := d.GetOk("relation_bgp_rs_best_path_ctrl_pol"); ok {
+			tfName := GetMOName(d.Get("relation_bgp_rs_best_path_ctrl_pol").(string))
+			if tfName != bgpRsBestPathCtrlPolData {
```
Comment (lhercot) [inline aci/resource_aci_bgpprotp.go:287]: I am confused on what you are trying to do here. tfName is not clear enough as a variable name for this relationship dn attribute.

File: aci/resource_aci_bgpprotp.go [hunk 5]
```diff
@@ -226,32 +276,47 @@ func resourceAciL3outBGPProtocolProfileRead(ctx context.Context, d *schema.Resou
...
 	bgpRsBgpNodeCtxPolData, err := aciClient.ReadRelationbgpRsBgpNodeCtxPolFromL3outBGPProtocolProfile(dn)
 	if err != nil {
 		log.Printf("[DEBUG] Error while reading relation bgpRsBgpNodeCtxPol %v", err)
 		d.Set("relation_bgp_rs_bgp_node_ctx_pol", "")
-
 	} else {
-		setRelationAttribute(d, "relation_bgp_rs_bgp_node_ctx_pol", bgpRsBgpNodeCtxPolData.(string))
+		if _, ok := d.GetOk("relation_bgp_rs_bgp_node_ctx_pol"); ok {
+			tfName := GetMOName(d.Get("relation_bgp_rs_bgp_node_ctx_pol").(string))
+			if tfName != bgpRsBgpNodeCtxPolData {
```
Comment (lhercot) [inline aci/resource_aci_bgpprotp.go:300]: Make tfName clearer.

**Review by lhercot — CHANGES_REQUESTED** [review]

File: aci/resource_aci_bgpprotp.go (hunk 2 above)
Comment (xinyuezhao) [reply aci/resource_aci_bgpprotp.go:50]: removed

File: aci/resource_aci_bgpprotp.go (hunk 3 above)
Comment (xinyuezhao) [reply aci/resource_aci_bgpprotp.go:166]: removed

File: aci/resource_aci_bgpprotp.go [hunk 6]
```diff
@@ -30,19 +30,26 @@ func resourceAciL3outBGPProtocolProfile() *schema.Resource {
...
+				Optional: true,
+				Computed: true,
+			},
 			"annotation": &schema.Schema{
 				Type:     schema.TypeString,
 				Optional: true,
 				Default:  "orchestrator:terraform",
 			},
+			"relation_bgp_rs_best_path_ctrl_pol": &schema.Schema{
+				Type: schema.TypeString,
```
Comment (lhercot) [inline aci/resource_aci_bgpprotp.go:50]: Why an empty space in the middle?

File: aci/resource_aci_bgpprotp.go [hunk 7]
```diff
@@ -138,49 +159,67 @@ func resourceAciL3outBGPProtocolProfileCreate(ctx context.Context, d *schema.Res
...
+
+	}
+
 	if relationTobgpRsBgpNodeCtxPol, ok := d.GetOk("relation_bgp_rs_bgp_node_ctx_pol"); ok {
 		relationParam := GetMOName(relationTobgpRsBgpNodeCtxPol.(string))
 		err = aciClient.CreateRelationbgpRsBgpNodeCtxPolFromL3outBGPProtocolProfile(bgpProtP.DistinguishedName, relationParam)
 		if err != nil {
 			return diag.FromErr(err)
 		}
+
```
Comment (lhercot) [inline aci/resource_aci_bgpprotp.go:177]: do we need this new line?

File: aci/resource_aci_bgpprotp.go [hunk 8]
```diff
@@ -226,32 +273,38 @@ func resourceAciL3outBGPProtocolProfileRead(ctx context.Context, d *schema.Resou
...
 		return nil
 	}
+
+	bgpRsBestPathCtrlPolData, err := aciClient.ReadRelationbgpRsBestPathCtrlPol(dn)
+	if err != nil {
+		log.Printf("[DEBUG] Error while reading relation bgpRsBestPathCtrlPol %v", err)
+		d.Set("relation_bgp_rs_best_path_ctrl_pol", "")
+	} else {
+		d.Set("relation_bgp_rs_best_path_ctrl_pol", bgpRsBestPathCtrlPolData)
+		log.Printf("[DEBUG] d.get relation_bgp_rs_best_path_ctrl_pol %v", d.Get("relation_bgp_rs_best_path_ctrl_pol"))
```
Comment (lhercot) [inline aci/resource_aci_bgpprotp.go:283]: do we still need this debug log?

**Review by lhercot — CHANGES_REQUESTED** [review]

**Review by lhercot — APPROVED** [review]
LGTM

**Pull request merged**